### PR TITLE
Issue053/sonar cards features

### PR DESCRIPTION
**Pull request by wisskirchenj**

### CODE REVIEW

File: server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java [hunk 1]
```diff
@@ -1,45 +1,50 @@
...
+)
+@JsonSubTypes({
+        @JsonSubTypes.Type(value = QuestionAndAnswer.class, name = "qna"),
+        @JsonSubTypes.Type(value = SingleChoiceQuiz.class, name = "scq"),
+        @JsonSubTypes.Type(value = MultipleChoiceQuiz.class, name = "mcq"),
+})
+public abstract sealed class Card permits QuestionAndAnswer, SingleChoiceQuiz, MultipleChoiceQuiz {
     @Id
-    @JsonDeserialize(using = MongoObjectIdConverter.class)
     @JsonAlias({"id", "_id"})
```
Comment (alfabravo2013) [inline server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java]: Now this alias can be safely removed.

File: server/src/main/java/org/hyperskill/community/flashcards/card/response/CardType.java [hunk 2]
```diff
@@ -1,7 +1,7 @@
 package org.hyperskill.community.flashcards.card.response;
 
 public enum CardType {
```
Comment (alfabravo2013) [inline server/src/main/java/org/hyperskill/community/flashcards/card/response/CardType.java:3]: I guess we can use it in `CardResponse` as the type of the `type` field?

File: server/src/main/java/org/hyperskill/community/flashcards/card/response/CardResponse.java [hunk 3]
```diff
@@ -2,30 +2,21 @@
...
-        this.createdAt = createdAt;
-        this.actions = actions;
-    }
+@SuperBuilder
+public abstract sealed class CardResponse permits QuestionAndAnswerDto, SingleChoiceQuizDto, MultipleChoiceQuizDto {
+    private final String id;
+    private final String title;
+    private final String question;
+    private final Set<String> tags;
+    private final String type;
```
Comment (alfabravo2013) [inline server/src/main/java/org/hyperskill/community/flashcards/card/response/CardResponse.java]: Can be of `CardType` type, right?

File: server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java [hunk 4]
```diff
@@ -1,45 +1,50 @@
...
  */
 @Getter
 @Setter
-@Document
 @ToString
-public sealed abstract class Card permits QuestionAndAnswer, SingleChoiceQuiz, MultipleChoiceQuiz {
+@NoArgsConstructor
+@SuperBuilder
+@Document
+@JsonTypeInfo(
```
Comment (alfabravo2013) [inline server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java:27]: Funny, but this is the only approach I didn't try. Was lured away by recommendations to create a converter and combine it with Jackson type hints. Can depend on the Spring Data version, of course.

**Review by alfabravo2013 — APPROVED** [review]

File: server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java (hunk 1 above)
Comment (wisskirchenj) [reply server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java]: yep - removed it

File: server/src/main/java/org/hyperskill/community/flashcards/card/response/CardResponse.java (hunk 3 above)
Comment (wisskirchenj) [reply server/src/main/java/org/hyperskill/community/flashcards/card/response/CardResponse.java]: yep - changed to CardType

File: server/src/main/java/org/hyperskill/community/flashcards/card/response/CardType.java (hunk 2 above)
Comment (wisskirchenj) [reply server/src/main/java/org/hyperskill/community/flashcards/card/response/CardType.java:3]: yes - done

**Review by alfabravo2013 — APPROVED** [review]

**Pull request merged**